app/main.py

```diff
@@ -1,44 +1,67 @@
+import asyncio
 import time
+from typing import Awaitable, Any
 
 from iot.devices import HueLightDevice, SmartSpeakerDevice, SmartToiletDevice
 from iot.message import Message, MessageType
 from iot.service import IOTService
 
 
-def main() -> None:
+async def run_sequence(*functions: Awaitable[Any]) -> None:
+    for function in functions:
+        await function
+
+
+async def run_parallel(*functions: Awaitable[Any]) -> None:
+    await asyncio.gather(*functions)
+
+
+async def main() -> None:
     # create an IOT service
     service = IOTService()
 
     # create and register a few devices
     hue_light = HueLightDevice()
     speaker = SmartSpeakerDevice()
     toilet = SmartToiletDevice()
-    hue_light_id = service.register_device(hue_light)
-    speaker_id = service.register_device(speaker)
-    toilet_id = service.register_device(toilet)
 
-    # create a few programs
-    wake_up_program = [
-        Message(hue_light_id, MessageType.SWITCH_ON),
-        Message(speaker_id, MessageType.SWITCH_ON),
-        Message(speaker_id, MessageType.PLAY_SONG, "Rick Astley - Never Gonna Give You Up"),
-    ]
+    hue_light_id, speaker_id, toilet_id = await asyncio.gather(
+        service.register_device(hue_light),
+        service.register_device(speaker),
+        service.register_device(toilet)
+    )
 
-    sleep_program = [
-        Message(hue_light_id, MessageType.SWITCH_OFF),
-        Message(speaker_id, MessageType.SWITCH_OFF),
-        Message(toilet_id, MessageType.FLUSH),
-        Message(toilet_id, MessageType.CLEAN),
-    ]
+    await run_sequence(
+        service.run_program(
+            [
+                Message(hue_light_id, MessageType.SWITCH_ON),
+                Message(hue_light_id, MessageType.SWITCH_OFF),
+                Message(speaker_id, MessageType.PLAY_SONG, "Rick Astley - Never Gonna Give You Up"),
+            ]
+        ),
+        run_parallel(
+            service.run_program(
+                [
+                    Message(hue_light_id, MessageType.SWITCH_OFF),
+                    Message(speaker_id, MessageType.SWITCH_OFF),
+                ]
+            ),
+            run_sequence(
+                service.run_program(
+                    [
+                        Message(toilet_id, MessageType.FLUSH),
+                        Message(toilet_id, MessageType.CLEAN),
+                    ]
+                )
+            )
+        )
+    )
```

Comment:
The way you're using `run_sequence` and `run_parallel` functions seems a bit complicated and may be hard to understand for others. Consider simplifying the code by using `asyncio.gather` and `asyncio.sleep` for running tasks in parallel and sequence respectively. Remember that clear and simple code is usually more maintainable.